### PR TITLE
[WIP] Added metric and alert for Uninstallation Stuck Due To PVC

### PR DESCRIPTION
**Pull request by bindrad**
Signed-off-by: bindrad <dbindra@redhat.com>

### CODE REVIEW

File: controllers/managedocs_controller.go [hunk 1]
```diff
@@ -1234,6 +1246,37 @@ func (r *ManagedOCSReconciler) reconcileOCSCSV() error {
...
+			return err
+		}
+
+		desired := templates.AdditionalPrometheusRule.DeepCopy()
+
+		for _, group := range desired.Spec.Groups {
+			if group.Name == "pvc-uninstall-stuck-alert" {
+				for _, rule := range group.Rules {
+					if rule.Alert == "UninstallStuckDueToPVC" {
+						rule.Labels["namespace"] = r.namespace
```
Comment (kesavanvt) [inline controllers/managedocs_controller.go:1263]: I see you are adding only the namespace here, any specific reason why we are filtering by group name and alertname?. Can we simply add namespace to all existing rules?

File: controllers/managedocs_controller.go [hunk 2]
```diff
@@ -1234,6 +1246,37 @@ func (r *ManagedOCSReconciler) reconcileOCSCSV() error {
...
+				}
+			}
+		}
+
+		r.additionalRule.Spec = desired.Spec
+
+		return nil
+	})
+	if err != nil {
+		return err
```
Comment (kesavanvt) [inline controllers/managedocs_controller.go:1274]: return meaningful error message

File: metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package metrics
```
Comment (kesavanvt) [inline metrics/metrics.go:1]: Add license information

File: templates/additionalPrometheusRule.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+*/
+
+package templates
+
+import (
+	promv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// This prometheus rule ensures that a DMS alert occurs during every prometheus scrape.
```
Comment (kesavanvt) [inline templates/additionalPrometheusRule.go:24]: DMS alert ?

**Comment by kesavanvt** [comment]
Modify template filename to lowercase 
`additionalPrometheusRule.go`  ->` additionalprometheusrule.go `

File: controllers/managedocs_controller.go (hunk 1 above)
Comment (bindrad) [reply controllers/managedocs_controller.go]: We only have two Prometheus rules in the deployer. First one is DMS and second one is this additional prometheus rule. DMS gets labeled in reconcileDMSPrometheusRule and AdditionalPrometheusRule get labeled in reconcileAdditionalPrometheusRule.

Comment (kesavanvt) [reply controllers/managedocs_controller.go]: I agree that both rules are labeled in different methods. My question was why do we check for `group.Name` to be `pvc-uninstall-stuck-alert` and `rule.Alert` to be `UninstallStuckDueToPVC`. Any reason on why we are doing this?. Why can't we just traverse through all the groups and alerts in `AdditionalPrometheusRule` and label with namespace?.

Comment (bindrad) [reply controllers/managedocs_controller.go]: Yes, I can do that. Anyways all the additionalPrometheusRule will need the namespace label

File: controllers/managedocs_controller.go [hunk 5]
```diff
@@ -83,6 +85,7 @@ const (
 	grafanaDatasourceSecretKey             = "prometheus.yaml"
 	k8sMetricsServiceMonitorAuthSecretName = "k8s-metrics-service-monitor-auth"
 	openshiftMonitoringNamespace           = "openshift-monitoring"
+	additionalRule                         = "additional-prometheus-rule"
```
Comment (kesavanvt) [inline controllers/managedocs_controller.go:88]:
```suggestion
	additionalRuleName                         = "additional-prometheus-rule"
```

File: controllers/managedocs_controller.go [hunk 6]
```diff
@@ -1149,6 +1163,29 @@ func (r *ManagedOCSReconciler) reconcileIngressNetworkPolicy() error {
...
 }
 
+func (r *ManagedOCSReconciler) reconcileAdditionalMetrics() error {
+	if r.checkUninstallCondition() {
+		metrics.ODFAddonPhase.With(
+			prometheus.Labels{
+				"status": "Uninstalling",
+			},
+		).Set(1)
+	} else if r.areComponentsReadyForUninstall() {
```
Comment (kesavanvt) [inline controllers/managedocs_controller.go:1173]: invoking this method is misleading though it satisfies the need, can we modify the methodname to `areComponentsReady`?

Comment (bindrad) [reply controllers/managedocs_controller.go:1173]: Yes, I'll change the method name

**Comment by bindrad** [comment]
deemed to be obsolete

**Pull request closed without merging**